File: go/content.go
package stem

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
)

// Built-in ContentPart type values from the crtx v0.1 enum.
const (
	PartTypeText       = "text"
	PartTypeToolCall   = "tool_call"
	PartTypeToolResult = "tool_result"
	PartTypeImage      = "image"
	PartTypeThinking   = "thinking"
)

// ContentPart is a discriminated union on Type. v0.1 variants:
// text, tool_call, tool_result, image, thinking. Extension parts
// MUST use a Type with the `x-` prefix; their fields round-trip
// verbatim via raw.
//
// Producers should construct ContentPart via the typed constructors
// (TextPart, ToolCallPart, ToolResultPart, etc.) or by populating
// Type plus the matching variant fields directly.
type ContentPart struct {
	// Type discriminates the variant. See PartType* constants.
	Type string `json:"type"`

	// Metadata is free-form and round-trips on every variant.
	Metadata map[string]any `json:"metadata,omitempty"`

	// text variant.
	Text string `json:"-"`

	// tool_call variant.
	CallID       string          `json:"-"`
	Name         string          `json:"-"`
	Input        json.RawMessage `json:"-"`
	ParentCallID string          `json:"-"`

	// tool_result variant.
	// (CallID reused.)
	Output          json.RawMessage `json:"-"`
	IsError         bool            `json:"-"`
	ChildEnvelopeID string          `json:"-"`

	// image variant.
	Mime string `json:"-"`
	Data string `json:"-"`
	URL  string `json:"-"`
	Alt  string `json:"-"`

	// thinking variant.
	// (Text reused.)
	Signature string `json:"-"`

	// raw preserves the entire JSON for extension parts (Type prefixed
	// "x-") plus any unknown fields on known parts. On extension parts
	// raw IS the canonical form; on known parts raw is nil after a
	// round-trip-clean unmarshal.
	raw json.RawMessage
}

// TextPart constructs a text ContentPart.
func TextPart(text string) ContentPart {
	return ContentPart{Type: PartTypeText, Text: text}
}

// ToolCallPart constructs a tool_call ContentPart. input may be a
// json.RawMessage already-encoded, or any JSON-marshalable value.
func ToolCallPart(callID, name string, input any) (ContentPart, error) {
	raw, err := toRawJSON(input)
	if err != nil {
		return ContentPart{}, fmt.Errorf("stem: tool_call input: %w", err)
	}
	return ContentPart{
		Type:   PartTypeToolCall,
		CallID: callID,
		Name:   name,
		Input:  raw,
	}, nil
}

// ToolResultPart constructs a tool_result ContentPart.
func ToolResultPart(callID string, output any, isError bool) (ContentPart, error) {
	raw, err := toRawJSON(output)
	if err != nil {
		return ContentPart{}, fmt.Errorf("stem: tool_result output: %w", err)
	}
	return ContentPart{
		Type:    PartTypeToolResult,
		CallID:  callID,
		Output:  raw,
		IsError: isError,
	}, nil
}

// IsExtension reports whether Type is an x-prefixed extension.
// Returns true for any "x-" prefix; structural validity of the
// trailing token is enforced separately by isValidExtensionType,
// invoked from Validate.
func (c ContentPart) IsExtension() bool {
	return strings.HasPrefix(c.Type, "x-")
}

// isValidMimeType reports whether s matches the crtx v0.1 image.mime
// regex ^[a-z]+/[a-zA-Z0-9.+-]+$. type token is lowercase letters
// only; subtype permits alphanumerics plus ".", "+", "-".
func isValidMimeType(s string) bool {
	slash := strings.IndexByte(s, '/')
	if slash < 1 || slash == len(s)-1 {
		return false
	}
	for _, r := range s[:slash] {
		if r < 'a' || r > 'z' {
			return false
		}
	}
	for _, r := range s[slash+1:] {
		switch {
		case r >= 'a' && r <= 'z':
		case r >= 'A' && r <= 'Z':
		case r >= '0' && r <= '9':
		case r == '.' || r == '+' || r == '-':
		default:
			return false
		}
	}
	return true
}

// isValidExtensionType reports whether s matches the crtx v0.1
// extension type regex ^x-[a-zA-Z0-9._-]+$. Requires at least one
// character after the "x-" prefix and restricts that character class
// to alphanumerics plus ".", "_", "-".
func isValidExtensionType(s string) bool {
	if len(s) < 3 || !strings.HasPrefix(s, "x-") {
		return false
	}
	for _, r := range s[2:] {
		switch {
		case r >= 'a' && r <= 'z':
		case r >= 'A' && r <= 'Z':
		case r >= '0' && r <= '9':
		case r == '.' || r == '_' || r == '-':
		default:
			return false
		}
	}
	return true
}

// clone deep-copies a ContentPart.
func (c ContentPart) clone() ContentPart {
	cp := c
	if c.Input != nil {
		cp.Input = append(json.RawMessage(nil), c.Input...)
	}
	if c.Output != nil {
		cp.Output = append(json.RawMessage(nil), c.Output...)
	}
	if c.raw != nil {
		cp.raw = append(json.RawMessage(nil), c.raw...)
	}
	if c.Metadata != nil {
		cp.Metadata = make(map[string]any, len(c.Metadata))
		for k, v := range c.Metadata {
			cp.Metadata[k] = v
		}
	}
	return cp
}

func toRawJSON(v any) (json.RawMessage, error) {
	if v == nil {
		return json.RawMessage(`null`), nil
	}
	if raw, ok := v.(json.RawMessage); ok {
		return append(json.RawMessage(nil), raw...), nil
	}
	if raw, ok := v.([]byte); ok {
		// Validate it parses.
		var probe any
		if err := json.Unmarshal(raw, &probe); err != nil {
			return nil, err
		}
		return append(json.RawMessage(nil), raw...), nil
	}
	return json.Marshal(v)
}

// MarshalJSON renders a ContentPart in crtx-canonical shape for its
// type discriminator. Extension parts emit their stored raw form.
func (c ContentPart) MarshalJSON() ([]byte, error) {
	if c.IsExtension() {
		if len(c.raw) == 0 {
			// Best-effort: emit just {"type": "..."} for an extension
			// part with no preserved body.
			return json.Marshal(struct {
				Type string `json:"type"`
			}{c.Type})
		}
		return c.raw, nil
	}

	switch c.Type {
	case PartTypeText:
		return json.Marshal(struct {
			Type     string         `json:"type"`
			Text     string         `json:"text"`
			Metadata map[string]any `json:"metadata,omitempty"`
		}{c.Type, c.Text, c.Metadata})

	case PartTypeToolCall:
		// Input is required by schema; ensure non-nil.
		input := c.Input
		if len(input) == 0 {
			input = json.RawMessage(`{}`)
		}
		return json.Marshal(struct {
			Type         string          `json:"type"`
			CallID       string          `json:"call_id"`
			Name         string          `json:"name"`
			Input        json.RawMessage `json:"input"`
			ParentCallID string          `json:"parent_call_id,omitempty"`
			Metadata     map[string]any  `json:"metadata,omitempty"`
		}{c.Type, c.CallID, c.Name, input, c.ParentCallID, c.Metadata})

	case PartTypeToolResult:
		output := c.Output
		if len(output) == 0 {
			output = json.RawMessage(`null`)
		}
		// Use a buffer to control field order + omit is_error /
		// child_envelope_id when unset.
		var buf bytes.Buffer
		buf.WriteString(`{"type":"tool_result","call_id":`)
		callIDBytes, err := json.Marshal(c.CallID)
		if err != nil {
			return nil, err
		}
		buf.Write(callIDBytes)
		buf.WriteString(`,"output":`)
		buf.Write(output)
		if c.IsError {
			buf.WriteString(`,"is_error":true`)
		}
		if c.ChildEnvelopeID != "" {
			childBytes, err := json.Marshal(c.ChildEnvelopeID)
			if err != nil {
				return nil, err
			}
			buf.WriteString(`,"child_envelope_id":`)
			buf.Write(childBytes)
		}
		if len(c.Metadata) > 0 {
			metaBytes, err := json.Marshal(c.Metadata)
			if err != nil {
				return nil, err
			}
			buf.WriteString(`,"metadata":`)
			buf.Write(metaBytes)
		}
		buf.WriteByte('}')
		return buf.Bytes(), nil

	case PartTypeImage:
		// Schema requires exactly one of data/url. Reject the producer
		// when neither or both are set; silent emission of invalid
		// JSON is a footgun callers cannot see until validation runs.
		hasData := c.Data != ""
		hasURL := c.URL != ""
		if hasData == hasURL {
			return nil, fmt.Errorf("stem: image part: exactly one of data/url required")
		}
		type imageOut struct {
			Type     string         `json:"type"`
			Mime     string         `json:"mime"`
			Data     string         `json:"data,omitempty"`
			URL      string         `json:"url,omitempty"`
			Alt      string         `json:"alt,omitempty"`
			Metadata map[string]any `json:"metadata,omitempty"`
		}
		return json.Marshal(imageOut{
			Type: c.Type, Mime: c.Mime, Data: c.Data, URL: c.URL,
			Alt: c.Alt, Metadata: c.Metadata,
		})

	case PartTypeThinking:
		return json.Marshal(struct {
			Type      string         `json:"type"`
			Text      string         `json:"text"`
			Signature string         `json:"signature,omitempty"`
			Metadata  map[string]any `json:"metadata,omitempty"`
		}{c.Type, c.Text, c.Signature, c.Metadata})

	default:
		return nil, fmt.Errorf("stem: unknown ContentPart type %q (extension parts MUST use x- prefix)", c.Type)
	}
}

// UnmarshalJSON parses a ContentPart, populating the typed fields for
// known variants and preserving the raw JSON for extension parts.
func (c *ContentPart) UnmarshalJSON(data []byte) error {
	// Discriminator probe.
	var probe struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(data, &probe); err != nil {
		return fmt.Errorf("stem: ContentPart: %w", err)
	}
	if probe.Type == "" {
		return fmt.Errorf("stem: ContentPart: missing type")
	}
	c.Type = probe.Type

	if strings.HasPrefix(probe.Type, "x-") {
		c.raw = append(json.RawMessage(nil), data...)
		return nil
	}

	switch probe.Type {
	case PartTypeText:
		var v struct {
			Text     string         `json:"text"`
			Metadata map[string]any `json:"metadata,omitempty"`
		}
		if err := json.Unmarshal(data, &v); err != nil {
			return fmt.Errorf("stem: text part: %w", err)
		}
		c.Text = v.Text
		c.Metadata = v.Metadata
		return nil

	case PartTypeToolCall:
		var v struct {
			CallID       string          `json:"call_id"`
			Name         string          `json:"name"`
			Input        json.RawMessage `json:"input"`
			ParentCallID string          `json:"parent_call_id,omitempty"`
			Metadata     map[string]any  `json:"metadata,omitempty"`
		}
		if err := json.Unmarshal(data, &v); err != nil {
			return fmt.Errorf("stem: tool_call part: %w", err)
		}
		c.CallID = v.CallID
		c.Name = v.Name
		c.Input = v.Input
		c.ParentCallID = v.ParentCallID
		c.Metadata = v.Metadata
		return nil

	case PartTypeToolResult:
		var v struct {
			CallID          string          `json:"call_id"`
			Output          json.RawMessage `json:"output"`
			IsError         bool            `json:"is_error,omitempty"`
			ChildEnvelopeID string          `json:"child_envelope_id,omitempty"`
			Metadata        map[string]any  `json:"metadata,omitempty"`
		}
		if err := json.Unmarshal(data, &v); err != nil {
			return fmt.Errorf("stem: tool_result part: %w", err)
		}
		c.CallID = v.CallID
		c.Output = v.Output
		c.IsError = v.IsError
		c.ChildEnvelopeID = v.ChildEnvelopeID
		c.Metadata = v.Metadata
		return nil

	case PartTypeImage:
		var v struct {
			Mime     string         `json:"mime"`
			Data     string         `json:"data,omitempty"`
			URL      string         `json:"url,omitempty"`
			Alt      string         `json:"alt,omitempty"`
			Metadata map[string]any `json:"metadata,omitempty"`
		}
		if err := json.Unmarshal(data, &v); err != nil {
			return fmt.Errorf("stem: image part: %w", err)
		}
		c.Mime = v.Mime
		c.Data = v.Data
		c.URL = v.URL
		c.Alt = v.Alt
		c.Metadata = v.Metadata
		return nil

	case PartTypeThinking:
		var v struct {
			Text      string         `json:"text"`
			Signature string         `json:"signature,omitempty"`
			Metadata  map[string]any `json:"metadata,omitempty"`
		}
		if err := json.Unmarshal(data, &v); err != nil {
			return fmt.Errorf("stem: thinking part: %w", err)
		}
		c.Text = v.Text
		c.Signature = v.Signature
		c.Metadata = v.Metadata
		return nil

	default:
		return fmt.Errorf("stem: ContentPart: unknown type %q (extension parts MUST use x- prefix)", probe.Type)
	}
}

File: rs/src/content.rs
//! `Role` enum + `ContentPart` discriminated union for crtx v0.1.

use serde::{Deserialize, Serialize};

/// Role identifies the author of a Turn. Mirrors the crtx v0.1 enum.
///
/// Unknown role strings fail decode with a serde error — crtx spec §4
/// requires implementations to reject the envelope rather than coerce
/// silently to a catch-all variant.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    User,
    Assistant,
    Tool,
    System,
    Developer,
}

impl Role {
    pub fn as_str(&self) -> &'static str {
        match self {
            Role::User => "user",
            Role::Assistant => "assistant",
            Role::Tool => "tool",
            Role::System => "system",
            Role::Developer => "developer",
        }
    }
}

/// A single payload unit inside a [`crate::Turn`].
///
/// `ContentPart` is a discriminated union on `type`. v0.1 variants:
/// `Text`, `ToolCall`, `ToolResult`, `Image`, `Thinking`. Anything
/// else round-trips verbatim through [`ContentPart::Extension`] —
/// producers MUST prefix the extension type with `x-` per spec.
///
/// Serialization is hand-rolled (see the `impl Serialize` /
/// `impl Deserialize` below) so the known variants enforce
/// `deny_unknown_fields` while extension parts round-trip with
/// arbitrary extra fields preserved.
#[derive(Debug, Clone, PartialEq)]
pub enum ContentPart {
    Text {
        text: String,
        metadata: Option<serde_json::Value>,
    },
    ToolCall {
        call_id: String,
        name: String,
        input: serde_json::Value,
        /// References an outer `tool_call.call_id` still open at this
        /// position when this call was issued as nested dispatch. See
        /// envelope.md §6.2.
        parent_call_id: Option<String>,
        metadata: Option<serde_json::Value>,
    },
    ToolResult {
        call_id: String,
        output: serde_json::Value,
        is_error: bool,
        /// Points at the dispatch-nest child Envelope's `id` when the
        /// call was executed out-of-band as a fresh session. See
        /// envelope.md §7.2.
        child_envelope_id: Option<String>,
        metadata: Option<serde_json::Value>,
    },
    Image {
        mime: String,
        data: Option<String>,
        url: Option<String>,
        alt: Option<String>,
        metadata: Option<serde_json::Value>,
    },
    Thinking {
        text: String,
        signature: Option<String>,
        metadata: Option<serde_json::Value>,
    },
    /// Catch-all for `x-*` extension parts. The full original JSON
    /// object is preserved verbatim under `raw`; the discriminator
    /// (`type`) is captured in `type_name`.
    Extension {
        type_name: String,
        raw: serde_json::Value,
    },
}

/// Convenience constructor for [`ContentPart::Text`].
impl ContentPart {
    pub fn text(text: impl Into<String>) -> Self {
        Self::Text {
            text: text.into(),
            metadata: None,
        }
    }

    pub fn tool_call(
        call_id: impl Into<String>,
        name: impl Into<String>,
        input: serde_json::Value,
    ) -> Self {
        Self::ToolCall {
            call_id: call_id.into(),
            name: name.into(),
            input,
            parent_call_id: None,
            metadata: None,
        }
    }

    /// Constructs a `tool_call` with `parent_call_id` set, marking it
    /// as nested dispatch within an outer open call. See envelope.md
    /// §6.2.
    pub fn tool_call_nested(
        call_id: impl Into<String>,
        name: impl Into<String>,
        input: serde_json::Value,
        parent_call_id: impl Into<String>,
    ) -> Self {
        Self::ToolCall {
            call_id: call_id.into(),
            name: name.into(),
            input,
            parent_call_id: Some(parent_call_id.into()),
            metadata: None,
        }
    }

    pub fn tool_result(
        call_id: impl Into<String>,
        output: serde_json::Value,
        is_error: bool,
    ) -> Self {
        Self::ToolResult {
            call_id: call_id.into(),
            output,
            is_error,
            child_envelope_id: None,
            metadata: None,
        }
    }

    /// Constructs a `tool_result` whose execution ran out-of-band as a
    /// dispatch-nest child Envelope; `child_envelope_id` lets consumers
    /// walk to the full child transcript. See envelope.md §7.2.
    pub fn tool_result_with_child(
        call_id: impl Into<String>,
        output: serde_json::Value,
        is_error: bool,
        child_envelope_id: impl Into<String>,
    ) -> Self {
        Self::ToolResult {
            call_id: call_id.into(),
            output,
            is_error,
            child_envelope_id: Some(child_envelope_id.into()),
            metadata: None,
        }
    }

    /// Discriminator string as it would appear on the wire.
    pub fn type_name(&self) -> &str {
        match self {
            Self::Text { .. } => "text",
            Self::ToolCall { .. } => "tool_call",
            Self::ToolResult { .. } => "tool_result",
            Self::Image { .. } => "image",
            Self::Thinking { .. } => "thinking",
            Self::Extension { type_name, .. } => type_name,
        }
    }

    /// True for `x-*` extension parts.
    pub fn is_extension(&self) -> bool {
        matches!(self, Self::Extension { .. })
    }
}

// --- Custom (de)serialization to honour deny_unknown_fields on known
// variants while letting x-* extensions round-trip with arbitrary
// extra fields. The default `#[serde(tag = "type")]` derive cannot
// express both simultaneously; the hand-rolled impls below split the
// known variants (strict) from the extension catchall (verbatim).

impl Serialize for ContentPart {
    fn serialize<S: serde::Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
        use serde::ser::SerializeMap;

        match self {
            Self::Text { text, metadata } => {
                let mut map = ser.serialize_map(None)?;
                map.serialize_entry("type", "text")?;
                map.serialize_entry("text", text)?;
                if let Some(m) = metadata {
                    map.serialize_entry("metadata", m)?;
                }
                map.end()
            }
            Self::ToolCall {
                call_id,
                name,
                input,
                parent_call_id,
                metadata,
            } => {
                let mut map = ser.serialize_map(None)?;
                map.serialize_entry("type", "tool_call")?;
                map.serialize_entry("call_id", call_id)?;
                map.serialize_entry("name", name)?;
                map.serialize_entry("input", input)?;
                if let Some(p) = parent_call_id {
                    map.serialize_entry("parent_call_id", p)?;
                }
                if let Some(m) = metadata {
                    map.serialize_entry("metadata", m)?;
                }
                map.end()
            }
            Self::ToolResult {
                call_id,
                output,
                is_error,
                child_envelope_id,
                metadata,
            } => {
                let mut map = ser.serialize_map(None)?;
                map.serialize_entry("type", "tool_result")?;
                map.serialize_entry("call_id", call_id)?;
                map.serialize_entry("output", output)?;
                if *is_error {
                    map.serialize_entry("is_error", &true)?;
                }
                if let Some(c) = child_envelope_id {
                    map.serialize_entry("child_envelope_id", c)?;
                }
                if let Some(m) = metadata {
                    map.serialize_entry("metadata", m)?;
                }
                map.end()
            }
            Self::Image {
                mime,
                data,
                url,
                alt,
                metadata,
            } => {
                let mut map = ser.serialize_map(None)?;
                map.serialize_entry("type", "image")?;
                map.serialize_entry("mime", mime)?;
                let has_data = data.is_some();
                let has_url = url.is_some();
                if has_data == has_url {
                    return Err(serde::ser::Error::custom(
                        "image part: exactly one of data/url required",
                    ));
                }
                if let Some(d) = data {
                    map.serialize_entry("data", d)?;
                }
                if let Some(u) = url {
                    map.serialize_entry("url", u)?;
                }
                if let Some(a) = alt {
                    map.serialize_entry("alt", a)?;
                }
                if let Some(m) = metadata {
                    map.serialize_entry("metadata", m)?;
                }
                map.end()
            }
            Self::Thinking {
                text,
                signature,
                metadata,
            } => {
                let mut map = ser.serialize_map(None)?;
                map.serialize_entry("type", "thinking")?;
                map.serialize_entry("text", text)?;
                if let Some(s) = signature {
                    map.serialize_entry("signature", s)?;
                }
                if let Some(m) = metadata {
                    map.serialize_entry("metadata", m)?;
                }
                map.end()
            }
            Self::Extension { raw, .. } => raw.serialize(ser),
        }
    }
}

impl<'de> Deserialize<'de> for ContentPart {
    fn deserialize<D: serde::Deserializer<'de>>(de: D) -> Result<Self, D::Error> {
        // First decode into a generic Value so we can branch on the
        // discriminator, then apply strict per-variant decoding for
        // known types and verbatim capture for x-* extensions.
        let value = serde_json::Value::deserialize(de)?;
        let obj = value
            .as_object()
            .ok_or_else(|| serde::de::Error::custom("ContentPart: not a JSON object"))?;
        let type_name = obj
            .get("type")
            .and_then(|v| v.as_str())
            .ok_or_else(|| serde::de::Error::custom("ContentPart: missing type"))?
            .to_owned();

        if type_name.starts_with("x-") {
            return Ok(Self::Extension {
                type_name,
                raw: value,
            });
        }

        // Strict-decode known variants: build a typed struct per shape
        // and reject unknown fields.
        match type_name.as_str() {
            "text" => {
                let v: TextWire =
                    serde_json::from_value(value).map_err(serde::de::Error::custom)?;
                Ok(Self::Text {
                    text: v.text,
                    metadata: v.metadata,
                })
            }
            "tool_call" => {
                let v: ToolCallWire =
                    serde_json::from_value(value).map_err(serde::de::Error::custom)?;
                Ok(Self::ToolCall {
                    call_id: v.call_id,
                    name: v.name,
                    input: v.input,
                    parent_call_id: v.parent_call_id,
                    metadata: v.metadata,
                })
            }
            "tool_result" => {
                let v: ToolResultWire =
                    serde_json::from_value(value).map_err(serde::de::Error::custom)?;
                Ok(Self::ToolResult {
                    call_id: v.call_id,
                    output: v.output,
                    is_error: v.is_error.unwrap_or(false),
                    child_envelope_id: v.child_envelope_id,
                    metadata: v.metadata,
                })
            }
            "image" => {
                let v: ImageWire =
                    serde_json::from_value(value).map_err(serde::de::Error::custom)?;
                Ok(Self::Image {
                    mime: v.mime,
                    data: v.data,
                    url: v.url,
                    alt: v.alt,
                    metadata: v.metadata,
                })
            }
            "thinking" => {
                let v: ThinkingWire =
                    serde_json::from_value(value).map_err(serde::de::Error::custom)?;
                Ok(Self::Thinking {
                    text: v.text,
                    signature: v.signature,
                    metadata: v.metadata,
                })
            }
            other => Err(serde::de::Error::custom(format!(
                "ContentPart: unknown type {other:?} (extension parts MUST use x- prefix)"
            ))),
        }
    }
}

// --- Per-variant wire structs used purely for strict decoding. -------

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct TextWire {
    #[serde(rename = "type")]
    _type: TypeText,
    text: String,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ToolCallWire {
    #[serde(rename = "type")]
    _type: TypeToolCall,
    call_id: String,
    name: String,
    input: serde_json::Value,
    #[serde(default)]
    parent_call_id: Option<String>,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ToolResultWire {
    #[serde(rename = "type")]
    _type: TypeToolResult,
    call_id: String,
    output: serde_json::Value,
    #[serde(default)]
    is_error: Option<bool>,
    #[serde(default)]
    child_envelope_id: Option<String>,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ImageWire {
    #[serde(rename = "type")]
    _type: TypeImage,
    mime: String,
    #[serde(default)]
    data: Option<String>,
    #[serde(default)]
    url: Option<String>,
    #[serde(default)]
    alt: Option<String>,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ThinkingWire {
    #[serde(rename = "type")]
    _type: TypeThinking,
    text: String,
    #[serde(default)]
    signature: Option<String>,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

// Marker enums whose only valid value is the matching discriminator
// string. Lets the per-variant struct above reject mis-typed parts.
macro_rules! type_marker {
    ($name:ident, $literal:literal) => {
        #[derive(Deserialize)]
        enum $name {
            #[serde(rename = $literal)]
            Tag,
        }
    };
}
type_marker!(TypeText, "text");
type_marker!(TypeToolCall, "tool_call");
type_marker!(TypeToolResult, "tool_result");
type_marker!(TypeImage, "image");
type_marker!(TypeThinking, "thinking");

/// Reports whether `s` matches the crtx v0.1 extension type regex
/// `^x-[a-zA-Z0-9._-]+$`. Public so [`crate::validate`] (and tests)
/// can apply the same rule.
pub fn is_valid_extension_type(s: &str) -> bool {
    if s.len() < 3 || !s.starts_with("x-") {
        return false;
    }
    s[2..]
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-')
}

/// Reports whether `s` matches the crtx v0.1 image.mime regex
/// `^[a-z]+/[a-zA-Z0-9.+-]+$`. type token is lowercase letters only;
/// subtype permits alphanumerics plus `.`, `+`, `-`.
pub fn is_valid_mime_type(s: &str) -> bool {
    let Some(slash) = s.find('/') else {
        return false;
    };
    if slash == 0 || slash == s.len() - 1 {
        return false;
    }
    let (typ, rest) = s.split_at(slash);
    let sub = &rest[1..];
    if !typ.chars().all(|c| c.is_ascii_lowercase()) {
        return false;
    }
    sub.chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '+' || c == '-')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unknown_role_fails_serde_decode() {
        // Per crtx spec §4 implementations MUST reject unknown roles
        // rather than coerce. The `#[serde(other)]` catch-all that used
        // to live here was removed; surface the decode error instead.
        let json = r#""wizard""#;
        let res: Result<Role, _> = serde_json::from_str(json);
        assert!(res.is_err());
    }

    #[test]
    fn extension_type_regex() {
        assert!(is_valid_extension_type("x-io.jadb.foo"));
        assert!(is_valid_extension_type("x-a"));
        assert!(!is_valid_extension_type("x-"));
        assert!(!is_valid_extension_type("x"));
        assert!(!is_valid_extension_type("x-bad space"));
        assert!(!is_valid_extension_type("custom"));
    }
}
